feat: Add point-in-sector utility to check for presence in room sectors

davidklaw opened the pull request:
This PR adds a new `IsPointInSector` function that lets you check if a specific point (by coordinates) is inside one or more sectors in a room. The function takes an array of IDs (sometimes called server IDs), so you can check multiple sectors at once. The main reason for adding this is to help catch players who hang around too long in guild hall door sectors, but it’s flexible enough to be useful in other situations too.

### Sector Geometry Loading on Startup
When the server starts up, it loads sector geometry from the room files. Each sector’s polygons get stored as arrays of vertex coordinates.

### IsPointInSector Coordinate Lookup Process
`IsPointInSector` converts Blakod coordinates to fine coordinates, grabs the room data, and loops through the sectors. It looks for sectors with IDs matching any of the provided user IDs, and for each match, uses ray-casting to see if the point is inside the sector’s polygon.

### Memory Usage Changes
**BEFORE:**
Memory Allocated: 30096561 bytes

**AFTER:**
Memory Allocated: 34324621 bytes

**DIFFERENCE:** 4,228,060 bytes / 4.22806 Megabytes

### Example: Guild Hall 11 (`guildh11`) (removed)
As a demo, I’ve updated `ghall.kod` and `guildh11.kod` to use this new function. To try it out: open the hall’s entrance door and stand in either of the two door sectors. Let the door close, and after a short delay, you’ll be teleported to the foyer. Modify `plDoorSectors` to test other scenarios, like an invalid sector id or a single integer:

```
plDoorSectors = [ MAINDOOR, MAINDOOR2 ]; // original
plDoorSectors = MAINDOOR; // single sector check
plDoorSectors = [ MAINDOOR, MAINDOOR2, "foo" ]; // invalid sector id, should still work
```

### Usage Examples
A single usage example was originally included included but removed in favor of a follow-up PR. The following is an example of how this could be used, though there's a bit more work needed to support it that can be seen in https://github.com/Meridian59/Meridian59/pull/1189/changes/a936b59e0027042d3d2fb2f92a75287cb9fb4809

```
// Check single user ID
if IsPointInSector(prmRoom, 
    Send(who,@GetRow), 
    Send(who,@GetCol), 
    Send(who,@GetFineRow), 
    Send(who,@GetFineCol), 
    DOOR_SECTOR_ID) 
{
    // Point is in sector for this user
}
```

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+static bool PointInPoly(int fx, int fy, const server_polygon &poly)
+{
+   if (poly.num_vertices < 3)
+      return False;
```
false

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+   if (poly.num_vertices < 3)
+      return False;
+
+   bool inside = False;
```
false

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+      int xj = poly.vertices_x[j], yj = poly.vertices_y[j];
+
+      int dy = yj - yi;
+      if (dy == 0)
```
dy unused... should be if (yj == yi) ?

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+                         int num_name_parms, parm_node name_parm_array[])
+{
+   val_type room_val, row_val, col_val, fr_val, fc_val, id_list_val;
+   std::vector<int> sector_ids;
```
Declaring a vector can allocate some memory, so it's better to declare it right before it's used.

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+         }
+         else
+         {
+            break;
```
This is an error, no?

davidklaw replied:
My intent was to handle the end of the list by simply breaking and moving on, using the sector ids already collected. But yes, it _could_ be an error, or it could simply be the end of the list.

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+   }
+   else
+   {
+      bprintf("C_PointInSector has bad 6th parameter tag, expected list or int\n");
```
A hard-coded number like "6th" is very likely to become outdated, making debugging impossible.  Better to tell the reader which parameter you mean.

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+   {
+      for (int i = 0; i < r->num_sectors; i++)
+      {
+         if (r->sectors[i].id != sector_id)
```
This embedded loop structure is odd.  Instead make a little loop that just finds the sector you're looking for and prints or returns an error if it's not found.  (Also, are the sectors in a room sorted by ID?  If so the lookup can be faster.)

davidklaw replied:
I'm fairly certain that sectors are not sorted by _server ID_, the relevant ID here. Maybe I should save sectors with server ids--the only ones this lookup cares about--separately from others?

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+         for (int p = 0; p < r->sectors[i].num_polygons; p++)
+         {
+            // First, check if the sector has polygons
+            if (r->sectors[i].polygons == NULL || p >= r->sectors[i].num_polygons)
```
This check seems unnecessary.

akirmse commented on blakserv/ccode.c:
```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
+               continue;
+            }
+
+            server_polygon *poly = &r->sectors[i].polygons[p];
```
This is a good place to use a reference:

const server_polygon &poly = r->sectors[i].polygons[p];

akirmse commented on blakserv/ccode.h:
```diff
@@ -245,7 +245,10 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
 		  int num_name_parms,parm_node name_parm_array[]);
 
-
+blak_int C_PointInSector(int object_id,local_var_type *local_vars,
```
I wonder if maybe "IsPointInSector" is a better name?  "PointInSector" suggests that perhaps it finds and returns a point in a sector.

akirmse commented on blakserv/ccode.c:
```diff
@@ -2366,6 +2372,204 @@ blak_int C_BuildString(int object_id, local_var_type *local_vars, int num_normal
+      return NIL;
+   }
+
+   if (fr_val.v.tag != TAG_INT)
```
Not checking fc_val?  Also, improve the error message ("fr" is cryptic).

akirmse commented on blakserv/ccode.c:
```diff
@@ -2366,6 +2372,204 @@ blak_int C_BuildString(int object_id, local_var_type *local_vars, int num_normal
+         }
+         else
+         {
+						break;
```
Fix indentation.

davidklaw commented:
Feedback has been addressed. Among the changes:
* Introduced vectors in room_type.
* Added validation for `fseek()`, `fread()`, and `readShort()` in `LoadSectorPolygons` to prevent crashes.
* Removed unreachable code in `C_IsPointInSector` where `TAG_INT` was checked after it had already been converted to `TAG_LIST`.

akirmse commented on blakserv/roofile.c:
```diff
@@ -17,122 +17,303 @@
+   room->cols = (short) temp;
+
+   room->grid = (unsigned char **) AllocateMemory(MALLOC_ID_ROOM, room->rows * sizeof(char *));
+   room->flags = (unsigned char **) AllocateMemory(MALLOC_ID_ROOM, room->rows * sizeof(char *));
```
Every error return after this leaks all of this memory, no?

akirmse commented on blakserv/roofile.c:
```diff
@@ -17,122 +17,303 @@
-   room->rows = static_cast<short>(temp);
+   // Read in subsection offsets
+   int node_off = readInt(infile);
+   readInt(infile); /* cwall off   */
```
The old code checked for errors on every read--the new code needs to do that too.

davidklaw commented:
Feedback addressed (and then some), and per your suggestion I added manual memory management methods to `memory`.

One additional change of note: Some of the `FreeMemory` calls in `BSPRooFileLoadServer` are using the wrong variable. They're using `i` instead of `j`. This PR corrects that, but I'll flip another PR for that alone in the morning if it's worthwhile.

https://github.com/Meridian59/Meridian59/blob/c555fbdd6cda1ab1be38ff122ba5f8999a62f8d7/blakserv/roofile.c#L89-L91

akirmse commented on blakserv/roomdata.c:
```diff
@@ -44,6 +44,23 @@ void InitRoomData()
+   total += room->sectors.capacity() * sizeof(server_sector);
+   
+   // Track each sector's polygon vector buffer
+   for (size_t i = 0; i < room->sectors.size(); i++)
```
You can just do something like 
```
for (auto &element : room->sectors) {
}
```
since you don't need the loop variable

akirmse commented on blakserv/roofile.c:
```diff
@@ -17,122 +17,535 @@
+   if (readInt(infile) == -1)
+   {
+      bprintf("BSPRooFileLoadServer: Failed to read sidedef_off in %s\n", fname);
+      fclose(infile);
```
I was hoping to save you all of these zillions of extra fclose calls and kind of excessive prints.  The file can be made to auto-close with various tricks (see RAII, or it can be done with a macro), but this change is already big enough as it is.

Also, bprintf is for Blakod-related errors.  These should all be eprintfs.

davidklaw replied:
I meant to mention that this was still on my list to address. It feels like something we should do and I'm happy to extract and submit the memory changes as a separate PR to reduce this some.

akirmse commented on blakserv/memory.c:
```diff
@@ -328,3 +328,18 @@ void * ResizeMemory(int malloc_id,void *ptr,int old_size,int new_size)
 #endif
 }
 
+void AddMemoryCount(int malloc_id, size_t size)
```
Could use a comment.

akirmse commented on blakserv/fileutil.h:
```diff
@@ -9,6 +9,38 @@
+   FILE *file;
+
+public:
+   FileGuard(FILE *f) : file(f) {}
```
Single argument constructors should be marked "explicit" to prevent automatic type coercion (weird C++ feature).

akirmse commented on blakserv/roomdata.c:
```diff
@@ -44,6 +44,22 @@ void InitRoomData()
    num_roomdata = 0;
 }
 
+static size_t CalculateRoomTypeMemory(const room_type *room)
```
Is this still used?

akirmse commented on blakserv/fileutil.h:
```diff
@@ -9,6 +9,38 @@
+ */
+class FileGuard
+{
+   FILE *file;
```
I should have been clearer--the default for classes is private, so the "file" data member is currently private.  You should add a private: section at the bottom and move "file" there.

davidklaw commented:
The latest commits redesign `readInt`, `readShort`, `readByte`, and `readCoord` helper functions to return `bool` and use output pointers instead of sentinel values, avoiding collisions with valid data. They also refactor the file macros to work with these changes. Other refactor work was to minimize comment real estate.

akirmse commented on blakserv/roofile.c:
```diff
@@ -19,119 +19,365 @@
+   else
+      *out = *((int *)buf);
+
+   // version ≤12 – bytes contain a plain little-endian int
```
This comment is in the wrong place.

akirmse commented on blakserv/roofile.c:
```diff
@@ -19,119 +19,365 @@
+   return fread(out, sizeof(short), 1, fd) == 1;
+}
+
+static bool readByte(FILE *fd, unsigned char *out)             /* 1-byte unsigned, returns int for error checking */
```
Comment isn't right.
